[quant][fx] Add root_node_getter in backend_config_dict

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73345

Summary:
For complex patterns we need to identify which node is the root, so that we can eliminate all other nodes and only preserve the root,
e.g. (torch.add, MatchAllNode, (torch.nn.ReLU, torch.nn.Conv2d)), we can preserve the torch.nn.Conv2d as root node, and remove other nodes.

Prevoiusly we assumed the root_node of a pattern is the "last node" of the pattern, computed by:
```
def default_root_node_getter(node_pattern):
    while not isinstance(node_pattern[-1], Node):
       node_pattern = node_pattern[-1]
       return node_pattern[-1]
```
This PR enables user configuration to define their own root_node_getter, that means we can define root_node for patterns like:
(torch.add, (torch.nn.ReLU, torch.nn.Conv2d), MatchAllNode)

Test Plan:
python test/test_quantize_fx.py TestFuseFx.test_root_node_getter

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34442193](https://our.internmc.facebook.com/intern/diff/D34442193)

jerryzh168 committed:
[quant][fx] Add root_node_getter in backend_config_dict

Summary:
For complex patterns we need to identify which node is the root, so that we can eliminate all other nodes and only preserve the root,
e.g. (torch.add, MatchAllNode, (torch.nn.ReLU, torch.nn.Conv2d)), we can preserve the torch.nn.Conv2d as root node, and remove other nodes.

Prevoiusly we assumed the root_node of a pattern is the "last node" of the pattern, computed by:
```
def default_root_node_getter(node_pattern):
    while not isinstance(node_pattern[-1], Node):
       node_pattern = node_pattern[-1]
       return node_pattern[-1]
```
This PR enables user configuration to define their own root_node_getter, that means we can define root_node for patterns like:
(torch.add, (torch.nn.ReLU, torch.nn.Conv2d), MatchAllNode)

Test Plan:
python test/test_quantize_fx.py TestFuseFx.test_root_node_getter

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34442193).

vkuzo commented:
I understand how root_node makes sense for sequential patterns.  Does it always make sense for an arbitrary pattern?  If not, is it an optional concept which doesn't always have to be defined?

vkuzo commented on torch/ao/quantization/fx/fuse.py:
```diff
@@ -75,10 +78,10 @@ def get_root_node(node_pattern):
-                # we want to make this more general to support more complex patterns
-                root_node = get_root_node(matched_node_pattern)  # type: ignore[index]
+                root_node_getter = fusion_pattern_to_root_node_getter.get(pattern, default_root_node_getter)
+                root_node = root_node_getter(matched_node_pattern)  # type: ignore[index]
```
should this have a check that root_node exists?

jerryzh168 replied:
we are inside the branch where we have a match I think, so we don't need to check if it exists, but we can have a validation that the root_node has the same type that is specified in the pattern, e.g. torch.nn.Conv2d

jerryzh168 commented:
> I understand how root_node makes sense for sequential patterns. Does it always make sense for an arbitrary pattern? If not, is it an optional concept which doesn't always have to be defined?

I feel it makes sense as well, it is basically the node that "remains" after the fusion, we are assuming we would have a single node that remains, which I feel is reasonable assumption in most of the cases

jerryzh168 commented on torch/ao/quantization/fx/fuse.py:
```diff
@@ -75,10 +78,10 @@ def get_root_node(node_pattern):
                 node_subpattern = None
             if maybe_last_node is node:
```
this means we have a successful match

vkuzo commented:
> I feel it makes sense as well, it is basically the node that "remains" after the fusion, we are assuming we would have a single node that remains, which I feel is reasonable assumption in most of the cases

Makes sense, might be good to explain that this is what we mean by "root node".  I think without explanation, people would assume that this is the first node of the original pattern, before replacement.  This is what I thought too, hence my questions above.

jerryzh168 commented:
Yes, will add documentations for backend_config_dict after we define it for fbgemm/qnnpack

jerryzh168 committed:
Update on "[quant][fx] Add root_node_getter in backend_config_dict"


Summary:
For complex patterns we need to identify which node is the root, so that we can eliminate all other nodes and only preserve the root,
e.g. (torch.add, MatchAllNode, (torch.nn.ReLU, torch.nn.Conv2d)), we can preserve the torch.nn.Conv2d as root node, and remove other nodes.

Prevoiusly we assumed the root_node of a pattern is the "last node" of the pattern, computed by:
```
def default_root_node_getter(node_pattern):
    while not isinstance(node_pattern[-1], Node):
       node_pattern = node_pattern[-1]
       return node_pattern[-1]
```
This PR enables user configuration to define their own root_node_getter, that means we can define root_node for patterns like:
(torch.add, (torch.nn.ReLU, torch.nn.Conv2d), MatchAllNode)

Test Plan:
python test/test_quantize_fx.py TestFuseFx.test_root_node_getter

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34442193](https://our.internmc.facebook.com/intern/diff/D34442193)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add root_node_getter in backend_config_dict"


Summary:
For complex patterns we need to identify which node is the root, so that we can eliminate all other nodes and only preserve the root,
e.g. (torch.add, MatchAllNode, (torch.nn.ReLU, torch.nn.Conv2d)), we can preserve the torch.nn.Conv2d as root node, and remove other nodes.

Prevoiusly we assumed the root_node of a pattern is the "last node" of the pattern, computed by:
```
def default_root_node_getter(node_pattern):
    while not isinstance(node_pattern[-1], Node):
       node_pattern = node_pattern[-1]
       return node_pattern[-1]
```
This PR enables user configuration to define their own root_node_getter, that means we can define root_node for patterns like:
(torch.add, (torch.nn.ReLU, torch.nn.Conv2d), MatchAllNode)

Test Plan:
python test/test_quantize_fx.py TestFuseFx.test_root_node_getter

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34442193](https://our.internmc.facebook.com/intern/diff/D34442193)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34442193).
